src/bin/lm_train_th.py

```diff
@@ -0,0 +1,437 @@
+                        help='Filename of validation label data (json)')
+    # LSTMLM training configuration
+    parser.add_argument('--batchsize', '-b', type=int, default=2048,
+                        help='Number of examples in each mini-batch')
+    parser.add_argument('--bproplen', '-l', type=int, default=35,
+                        help='Number of words in each mini-batch '
+                             '(= length of truncated BPTT)')
+    parser.add_argument('--epoch', '-e', type=int, default=20,
+                        help='Number of sweeps over the dataset to train')
+    parser.add_argument('--gradclip', '-c', type=float, default=5,
+                        help='Gradient norm threshold to clip')
+    parser.add_argument('--unit', '-u', type=int, default=650,
+                        help='Number of LSTM units in each layer')
+    args = parser.parse_args()
+
+    # logging info
+    if args.verbose > 0:
+        logging.basicConfig(
+            level=logging.INFO, format='%(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s')
+    else:
+        logging.basicConfig(
+            level=logging.WARN, format='%(asctime)s (%(module)s:%(lineno)d) %(levelname)s: %(message)s')
+        logging.warning('Skip DEBUG/INFO messages')
+
+    # display PYTHONPATH
+    logging.info('python path = ' + os.environ['PYTHONPATH'])
+
+    # display torch version
+    logging.info('torch version = ' + torch.__version__)
+
+    # seed setting (chainer seed may not need it)
+    nseed = args.seed
+    random.seed(nseed)
+    np.random.seed(nseed)
+    torch.manual_seed(nseed)
+    logging.info('torch seed = ' + str(nseed))
+
+    # debug mode setting
+    # 0 would be fastest, but 1 seems to be reasonable
+    # by considering reproducability
+    # use determinisitic computation or not
+    if args.debugmode < 1:
+        torch.backends.cudnn.deterministic = False
+        logging.info('torch cudnn deterministic is disabled')
+    else:
+        torch.backends.cudnn.deterministic = True
+
+    # load dictionary for debug log
+    with open(args.dict, 'rb') as f:
+        dictionary = f.readlines()
+    char_list = [entry.decode('utf-8').split(' ')[0] for entry in dictionary]
+    char_list.insert(0, '<blank>')
+    char_list.append('<eos>')
+    char_list_dict = {x: i for i, x in enumerate(char_list)}
+
+    # check cuda and cudnn availability
+    if not torch.cuda.is_available():
+        logging.warning('cuda is not available')
+
+    def evaluate(model, iter, bproplen=100):
+        # Evaluation routine to be used for validation and test.
+        model.predictor.eval()
+        state = {
+            'c1': model.predictor.zero_state(args.batchsize),
+            'h1': model.predictor.zero_state(args.batchsize),
+            'c2': model.predictor.zero_state(args.batchsize),
+            'h2': model.predictor.zero_state(args.batchsize)
+        }
+        if args.gpu >= 0:
+            for key in state.iterkeys():
```

Comment:
use `keys` instead of `iterkeys` that is unavailable in python3 and not memory consuming in this case